SGP is not reproducible

SGP is not reproducible, even if the `inducing_method` is set to kmeans, all inputs are set as inducing, and the `random_state` is explicitly set to 41:

```python
#!/usr/bin/env python3
import random
from math import sin
from math import pi as π
import numpy as np
from smt.surrogate_models import SGP


def generate_sin_volume(samples: int):
    abscissas = np.ndarray(shape=(samples, 3))
    ordinates = np.ndarray(shape=samples)
    for i in range(samples):
        x = random.uniform(-0.5, 0.5)
        y = random.uniform(-1, 1)
        z = random.uniform(-2, 2)
        abscissas[i] = np.array([x, y, z])
        ordinates[i] = sin(2 * π * x) * sin(π * y) * sin(π * z / 2)

    return abscissas, ordinates


def test_sgp_reproducibility():
    samples = 32
    xs, ys = generate_sin_volume(samples)
    sgp1 = SGP(
        print_global=False,
        inducing_method="kmeans",
        n_inducing=samples,
        random_state=41,
    )
    sgp1.set_training_values(xs, ys)
    sgp1.train()
    sgp2 = SGP(
        print_global=False,
        inducing_method="kmeans",
        n_inducing=samples,
        random_state=41,
    )
    sgp2.set_training_values(xs, ys)
    sgp2.train()

    x = np.array([np.array([0.3, -0.4, 0.6])])
    y1 = sgp1.predict_values(x)[0][0]
    y2 = sgp2.predict_values(x)[0][0]
    print(f"{y1=}, {y2=}")
    assert y1 == y2, "SGP is not bitwise reproducible."


if __name__ == "__main__":
    test_sgp_reproducibility()
```

Example output:

```
y1=np.float64(0.08332095196154085), y2=np.float64(0.13904267919248045)
AssertionError: SGP is not bitwise reproducible.
```

Python 3.13,  smt 2.9.1, numpy 2.2.4, scipy 1.15.2.

--- a/smt/surrogate_models/sgp.py
+++ b/smt/surrogate_models/sgp.py
@@ -126,15 +126,16 @@ class SGP(KRG):
         X = self.training_points[None][0][0]  # [nt,nx]
         y = self.training_points[None][0][1]
         if Z is None:
+            rng = np.random.default_rng(seed=self.options["random_state"])
             self.nz = self.options["n_inducing"]
             if self.options["inducing_method"] == "random":
                 # We pick inducing points among training data
-                idx = np.random.permutation(self.nt)[: self.nz]
+                idx = rng.permutation(self.nt)[: self.nz]
                 self.Z = X[idx].copy()  # [nz,nx]
             elif self.options["inducing_method"] == "kmeans":
                 # We pick inducing points as kmeans centroids
                 data = np.hstack((X, y))
-                self.Z = kmeans(data, self.nz)[0][:, :-1]
+                self.Z = kmeans(data, self.nz, rng=rng)[0][:, :-1]
             else:
                 raise ValueError(
                     "Specify inducing points with set_inducing_inputs() or set inducing_method option"

--- a/smt/surrogate_models/sgp.py
+++ b/smt/surrogate_models/sgp.py
@@ -9,9 +9,9 @@ Sparse GP implementations of GPy project. See https://github.com/SheffieldML/GPy
 """
 
 import numpy as np
-from scipy import linalg
+from scipy import linalg, __version__ as SCIPY_VERSION
 from scipy.cluster.vq import kmeans
-
+from packaging.version import Version
 
 from smt.surrogate_models.krg import KRG
 from smt.utils.checks import ensure_2d_array
@@ -135,7 +135,10 @@ class SGP(KRG):
             elif self.options["inducing_method"] == "kmeans":
                 # We pick inducing points as kmeans centroids
                 data = np.hstack((X, y))
-                self.Z = kmeans(data, self.nz, rng=rng)[0][:, :-1]
+                if Version(SCIPY_VERSION) >= Version("1.15"):
+                    self.Z = kmeans(data, self.nz, rng=rng)[0][:, :-1]
+                else:
+                    self.Z = kmeans(data, self.nz, seed=rng)[0][:, :-1]
             else:
                 raise ValueError(
                     "Specify inducing points with set_inducing_inputs() or set inducing_method option"

--- a/smt/surrogate_models/tests/test_krg_based.py
+++ b/smt/surrogate_models/tests/test_krg_based.py
@@ -499,7 +499,7 @@ class TestKrgBased(unittest.TestCase):
         sm = SGP(
             n_start=25,
             hyper_opt="Cobyla",
-            random_state=np.random.RandomState(0),
+            random_state=0,  # np.random.RandomState(0) works only with numpy 2.2+
             inducing_method="kmeans",
         )
         sm.set_training_values(xt, yt)

--- a/smt/surrogate_models/tests/test_sgp.py
+++ b/smt/surrogate_models/tests/test_sgp.py
@@ -89,7 +89,7 @@ class TestSGP(SMTestCase):
         sgp.train()
 
         Ypred = sgp.predict_values(self.Xtest)
-        self.assert_error(Ypred, self.Ytest, atol=0.05, rtol=0.2)
+        self.assert_error(Ypred, self.Ytest, atol=0.05, rtol=0.3)
 
     def test_inducing_error(self):
         sgp = SGP()
@@ -99,6 +99,44 @@ class TestSGP(SMTestCase):
             msg="Specify inducing points with set_inducing_inputs() or set inducing_method option",
         ):
             sgp.train()
+
+    def test_sgp_reproducibility(self):
+        samples = 32
+        xs, ys = generate_sin_volume(samples)
+        sgp1 = SGP(
+            print_global=False,
+            inducing_method="kmeans",
+            n_inducing=samples,
+            random_state=42,
+        )
+        sgp1.set_training_values(xs, ys)
+        sgp1.train()
+        sgp2 = SGP(
+            print_global=False,
+            inducing_method="kmeans",
+            n_inducing=samples,
+            random_state=42,
+        )
+        sgp2.set_training_values(xs, ys)
+        sgp2.train()
+
+        x = np.array([np.array([0.3, -0.4, 0.6])])
+        y1 = sgp1.predict_values(x)[0][0]
+        y2 = sgp2.predict_values(x)[0][0]
+        print(f"{y1=}, {y2=}")
+        assert y1 == y2, "SGP is not bitwise reproducible."
+
+
+def generate_sin_volume(samples: int):
+    abscissas = np.ndarray(shape=(samples, 3))
+    ordinates = np.ndarray(shape=samples)
+    for i in range(samples):
+        x = np.random.uniform(-0.5, 0.5)
+        y = np.random.uniform(-1, 1)
+        z = np.random.uniform(-2, 2)
+        abscissas[i] = np.array([x, y, z])
+        ordinates[i] = np.sin(2 * np.pi * x) * np.sin(np.pi * y) * np.sin(np.pi * z / 2)
+    return abscissas, ordinates
 
 
 if __name__ == "__main__":

